config/http_config.go

```diff
@@ -785,54 +815,69 @@ func NewTLSRoundTripper(
 		return nil, err
 	}
 	t.rt = rt
-	_, t.hashCAFile, err = t.getCAWithHash()
+	_, t.hashCAFile, t.hashCertFile, t.hashKeyFile, err = t.getTLSFilesWithHash()
 	if err != nil {
 		return nil, err
 	}
 
 	return t, nil
 }
 
-func (t *tlsRoundTripper) getCAWithHash() ([]byte, []byte, error) {
-	b, err := readCAFile(t.caFile)
+func (t *tlsRoundTripper) getTLSFilesWithHash() ([]byte, []byte, []byte, []byte, error) {
+	b1, err := readCAFile(t.caFile)
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, nil, nil, err
 	}
-	h := sha256.Sum256(b)
-	return b, h[:], nil
+	h1 := sha256.Sum256(b1)
 
+	b2, b3, err := readCertAndKey(t.certFile, t.keyFile)
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+	h2, h3 := sha256.Sum256(b2), sha256.Sum256(b3)
+
+	return b1, h1[:], h2[:], h3[:], nil
 }
 
 // RoundTrip implements the http.RoundTrip interface.
 func (t *tlsRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
-	b, h, err := t.getCAWithHash()
+	caData, caHash, certHash, keyHash, err := t.getTLSFilesWithHash()
 	if err != nil {
 		return nil, err
 	}
 
 	t.mtx.RLock()
-	equal := bytes.Equal(h[:], t.hashCAFile)
+	caEqual := bytes.Equal(caHash[:], t.hashCAFile)
+	certKeyEqual := bytes.Equal(certHash[:], t.hashCertFile) && bytes.Equal(keyHash[:], t.hashKeyFile)
 	rt := t.rt
 	t.mtx.RUnlock()
-	if equal {
+	if caEqual && certKeyEqual {
 		// The CA cert hasn't changed, use the existing RoundTripper.
 		return rt.RoundTrip(req)
 	}
 
-	// Create a new RoundTripper.
-	tlsConfig := t.tlsConfig.Clone()
-	if !updateRootCA(tlsConfig, b) {
-		return nil, fmt.Errorf("unable to use specified CA cert %s", t.caFile)
-	}
-	rt, err = t.newRT(tlsConfig)
-	if err != nil {
-		return nil, err
+	// The cert and key files are read separately by the client
+	// using GetClientCertificate, therefore the RoundTripper
+	// doesn't need to be updated if only they are changed.
```

Comment:
It actually does, otherwise it would not recreate new connections. t.CloseIdleConnections() is useless if other scrapes use the same http client and are in progress.